core/backoff/retry.go

```diff
@@ -70,66 +71,89 @@ type Strategy struct {
 //	    MaxRetries: 5,
 //	    Delay: 100,
 //	    Exponential: true,
-//	    MaxDuration: 10 * time.Second,
+//	    MaxWait: 10 * time.Second,
 //	})
 //	if err != nil {
 //	    fmt.Println("Failed to connect to database after retries:", err)
 //	}
 func Retry(scope string, f func() error, s Strategy) error {
+	cid := crypto.Id()
+
 	s = withDefaults(s)
 	var err error
 
+	log.TraceLn(&cid, "Retry: starting retry loop")
+
 	for i := 0; i <= int(s.MaxRetries); i++ {
 		err = f()
+
+		log.TraceLn(&cid, "Retry: executed the function")
+
 		if err == nil {
+			log.TraceLn(&cid, "Retry: success")
 			return nil
 		}
 
-		var retryDelay time.Duration
-		// if exponential backoff is enabled then delay increases exponentially
+		var multiplier float64 = 1
+
+		// if exponential backoff is enabled then delay increases exponentially:
 		if s.Exponential {
-			// Calculate the delay for the current attempt. The delay is
-			// 2^i seconds.
-			delay := time.Duration(math.Pow(2, float64(i))) * time.Second
-			// Some randomness to avoid the thundering herd problem.
-			d := int(s.Delay)
-			delay += time.Duration(rand.Intn(d)) * time.Millisecond
-			if delay > s.MaxDuration {
-				delay = s.MaxDuration
-			}
-
-			retryDelay = delay
-		} else { // otherwise delay is constant for all retries
-			retryDelay = s.Delay * time.Millisecond
+			multiplier = math.Pow(2, float64(i))
+		}
+
+		sDelayMs := s.Delay.Milliseconds()
+		if sDelayMs == 0 {
+			sDelayMs = 10
+		}
+
+		delayMs := multiplier * float64(sDelayMs)
+		delay := time.Duration(delayMs) * time.Millisecond
+
+		// Some randomness to avoid the thundering herd problem.
+		jitter := rand.Intn(int(sDelayMs))
+		delay += time.Duration(jitter) * time.Millisecond
+		if delay > s.MaxWait {
+			delay = s.MaxWait
 		}
 
-		time.Sleep(retryDelay)
-		_, _ = fmt.Printf(
-			"Retrying after %d ms for the scope '%s' -- attempt %d of %d",
-			retryDelay, scope, i+1, s.MaxRetries+1,
+		log.TraceLn(&cid, "Retry: will sleep:", delay)
```

Comment:
using the standard logging library, so that things can be managed/forwarded/aggregated from a central configuration (later).